### ISSUE TITLE
Not stepdown master when there aren't enough master-eligeble nodes

### ISSUE THREAD

**Issue by johtani** [elastic/elasticsearch#22108]
**Elasticsearch version**: 5.x

**Description of the problem including expected versus actual behavior**:

Not stepdown master when there aren't enough master-eligeble nodes.
It should be no master cluster, but the cluster have master node.
In 2.4.2, the master node stepdown and the cluster has no master.

**Steps to reproduce**:
 1. Set `discovery.zen.minimum_master_nodes: 1`
 2. Start 2 elasticsearch nodes, es1 and es2
 3. Update cluster settings
```
curl -XPUT localhost:9200/_cluster/settings -d '{
"persistent": {
    "discovery.zen.minimum_master_nodes": 2
  }
}'
```
4. Stop es2, then we can not get _cluster/settings because there is no master.
5. Stop es1 and start es1
6. Get _cluster/settings. we can still get _cluster/settings and also get _cat/master

**Provide logs (if relevant)**:

```
[2016-12-12T15:50:11,387][INFO ][o.e.h.HttpServer         ] [VQmoLN6] publish_address {127.0.0.1:9200}, bound_addresses {[fe80::1]:9200}, {[::1]:9200}, {127.0.0.1:9200}
[2016-12-12T15:50:11,387][INFO ][o.e.n.Node               ] [VQmoLN6] started
[2016-12-12T15:50:11,392][WARN ][o.e.c.s.ClusterSettings  ] [VQmoLN6] failed to prepareCommit settings for [Updater for: {
  "key" : "discovery.zen.minimum_master_nodes",
  "properties" : [
    "Dynamic",
    "NodeScope"
  ],
  "is_group_setting" : false,
  "default" : "-1"
}]
java.lang.IllegalArgumentException: illegal value can't update [discovery.zen.minimum_master_nodes] from [1] to [2]
	at org.elasticsearch.common.settings.Setting$Updater.getValue(Setting.java:478) ~[elasticsearch-5.1.1.jar:5.1.1]
	at org.elasticsearch.common.settings.AbstractScopedSettings$SettingUpdater.updater(AbstractScopedSettings.java:329) ~[elasticsearch-5.1.1.jar:5.1.1]
	at org.elasticsearch.common.settings.AbstractScopedSettings.applySettings(AbstractScopedSettings.java:165) [elasticsearch-5.1.1.jar:5.1.1]
	at org.elasticsearch.cluster.service.ClusterService.runTasksForExecutor(ClusterService.java:729) [elasticsearch-5.1.1.jar:5.1.1]
	at org.elasticsearch.cluster.service.ClusterService$UpdateTask.run(ClusterService.java:920) [elasticsearch-5.1.1.jar:5.1.1]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:458) [elasticsearch-5.1.1.jar:5.1.1]
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:238) [elasticsearch-5.1.1.jar:5.1.1]
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:201) [elasticsearch-5.1.1.jar:5.1.1]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [?:1.8.0_91]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [?:1.8.0_91]
	at java.lang.Thread.run(Thread.java:745) [?:1.8.0_91]
Caused by: java.lang.IllegalArgumentException: cannot set discovery.zen.minimum_master_nodes to more than the current master nodes count [1]
	at org.elasticsearch.discovery.zen.ZenDiscovery.lambda$new$2(ZenDiscovery.java:168) ~[elasticsearch-5.1.1.jar:5.1.1]
	at org.elasticsearch.common.settings.Setting$Updater.getValue(Setting.java:476) ~[elasticsearch-5.1.1.jar:5.1.1]
	... 10 more
[2016-12-12T15:50:11,401][WARN ][o.e.c.s.ClusterSettings  ] [VQmoLN6] failed to apply settings
java.lang.IllegalArgumentException: illegal value can't update [discovery.zen.minimum_master_nodes] from [1] to [2]
	at org.elasticsearch.common.settings.Setting$Updater.getValue(Setting.java:478) ~[elasticsearch-5.1.1.jar:5.1.1]
	at org.elasticsearch.common.settings.AbstractScopedSettings$SettingUpdater.updater(AbstractScopedSettings.java:329) ~[elasticsearch-5.1.1.jar:5.1.1]
	at org.elasticsearch.common.settings.AbstractScopedSettings.applySettings(AbstractScopedSettings.java:165) [elasticsearch-5.1.1.jar:5.1.1]
	at org.elasticsearch.cluster.service.ClusterService.runTasksForExecutor(ClusterService.java:729) [elasticsearch-5.1.1.jar:5.1.1]
	at org.elasticsearch.cluster.service.ClusterService$UpdateTask.run(ClusterService.java:920) [elasticsearch-5.1.1.jar:5.1.1]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:458) [elasticsearch-5.1.1.jar:5.1.1]
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:238) [elasticsearch-5.1.1.jar:5.1.1]
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:201) [elasticsearch-5.1.1.jar:5.1.1]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [?:1.8.0_91]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [?:1.8.0_91]
	at java.lang.Thread.run(Thread.java:745) [?:1.8.0_91]
Caused by: java.lang.IllegalArgumentException: cannot set discovery.zen.minimum_master_nodes to more than the current master nodes count [1]
	at org.elasticsearch.discovery.zen.ZenDiscovery.lambda$new$2(ZenDiscovery.java:168) ~[elasticsearch-5.1.1.jar:5.1.1]
	at org.elasticsearch.common.settings.Setting$Updater.getValue(Setting.java:476) ~[elasticsearch-5.1.1.jar:5.1.1]
	... 10 more
[2016-12-12T15:50:11,402][WARN ][o.e.c.s.ClusterService   ] [VQmoLN6] failed to apply cluster settings
java.lang.IllegalArgumentException: illegal value can't update [discovery.zen.minimum_master_nodes] from [1] to [2]
	at org.elasticsearch.common.settings.Setting$Updater.getValue(Setting.java:478) ~[elasticsearch-5.1.1.jar:5.1.1]
	at org.elasticsearch.common.settings.AbstractScopedSettings$SettingUpdater.updater(AbstractScopedSettings.java:329) ~[elasticsearch-5.1.1.jar:5.1.1]
	at org.elasticsearch.common.settings.AbstractScopedSettings.applySettings(AbstractScopedSettings.java:165) ~[elasticsearch-5.1.1.jar:5.1.1]
	at org.elasticsearch.cluster.service.ClusterService.runTasksForExecutor(ClusterService.java:729) [elasticsearch-5.1.1.jar:5.1.1]
	at org.elasticsearch.cluster.service.ClusterService$UpdateTask.run(ClusterService.java:920) [elasticsearch-5.1.1.jar:5.1.1]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:458) [elasticsearch-5.1.1.jar:5.1.1]
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:238) [elasticsearch-5.1.1.jar:5.1.1]
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:201) [elasticsearch-5.1.1.jar:5.1.1]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [?:1.8.0_91]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [?:1.8.0_91]
	at java.lang.Thread.run(Thread.java:745) [?:1.8.0_91]
Caused by: java.lang.IllegalArgumentException: cannot set discovery.zen.minimum_master_nodes to more than the current master nodes count [1]
	at org.elasticsearch.discovery.zen.ZenDiscovery.lambda$new$2(ZenDiscovery.java:168) ~[elasticsearch-5.1.1.jar:5.1.1]
	at org.elasticsearch.common.settings.Setting$Updater.getValue(Setting.java:476) ~[elasticsearch-5.1.1.jar:5.1.1]
	... 10 more
[2016-12-12T15:50:11,407][INFO ][o.e.g.GatewayService     ] [VQmoLN6] recovered [0] indices into cluster_state
```

**Comment by ywelsch** [comment]
Interesting find. This shows a discrepancy between node settings and persistent (cluster-level) settings. When es1 is restarted it uses its node-local setting of min_master_nodes (which is 1) to publish the initial cluster state (that's the cluster state that establishes es1 as master). It also does so for the second cluster state where the state is recovered (only the recovered state has the information min_master_nodes = 2). That state will not allow that the settings be applied on the local node though, as there is a check in place that prevents updating min_master_nodes to a value that is higher than the current number of master-eligible nodes. The node thus continues working as if it had min_master_nodes = 1 but is unable to apply settings locally 👎 .

I see a few options to address this:
1) Remove the check that prevents min_master_nodes to be set to a value that is higher than the current number of master-eligible nodes.
2) Don't allow the recovered state to be published.
3) make `min_master_nodes` a non-dynamic setting so that it can only be set at node-level.

What are the advantages / disadvantages of these solutions?

1) has the downside that only after publishing the state to itself will the node detect that min_master_nodes is set to less than a quorum and step down as master, i.e. there is a short time where the node will still appear as fully-recovered master (but at least it takes corrective action).
2) has a similar issue (although less bad). It makes the master switch between master-election and `state not recovered / initialized` mode as it can become master but not publish the recovered state. If, in the future, we don't have a separate step for recovering the state but tie this into master-election, this would be my preferred method.
3) eliminates the discrepancy between node and cluster-level but puts an additional burden on the user to make sure that every master-eligible node has the correct settings (i.e. that the node settings do not diverge among the master-eligible nodes). Also, BWC...

As a short-term fix, 1) seems to be the simplest solution. WDYT @bleskes @jasontedor?

**Comment by s1monw** [comment]
@ywelsch can we in this particular case check if the cluster state we use to get the number of master nodes in the cluster has the same value as the one we are trying to apply and then skip the check and apply? I think for this particular situation this is enough?

**Comment by ywelsch** [comment]
The check is called by the settings update validator, which is called in two contexts:
1) when the user submits a settings change and master validates the setting (dryRun).
2) when the cluster service applies a cluster state update with the updated setting (that was published and validated before).

If I understand your suggestion, you would like to distinguish between those 2 situations and only check the property in the first case. These two situations can be distinguished currently by looking at the cluster state available in `ClusterService.state()`. In case 1), the cluster state will be the one before applying this change, and in case 2), it will be the one after applying this change. With PR #21817, this changes however, and the 2 situations become indistinguishable in that regard (I've made a related comment as well on that PR regarding this aspect).

**Comment by DaveCTurner** [comment]
Closing this as it is superseded by the changes we have planned in the Zen 2 project.

**Comment by thomassuckow** [comment]
What is the status of Zen2? From what i can tell it looks like it is part of 7.0 and a new cluster.initial_master_nodes

**Comment by bleskes** [comment]
@thomassuckow zen2 will indeed be released as part of 7.0

**Comment by VGerris** [comment]
hi, I have a similar issue in Kubernetes when trying to scale down. Is there any way to fix this?

**Comment by VGerris** [comment]
fixed eventually by using:
curl -XPUT http://localhost:9200/_cluster/settings -d '{ "persistent" :{"discovery.zen.minimum_master_nodes" : 1 }}' -H 'Content-Type: application/json' om the working host and then back to 2 because using count 3.
